File: RhinoScriptCommand.cs
using System;
using System.IO;
using System.Security;
using System.Security.Cryptography;
using System.Reflection;
using System.Runtime.InteropServices;
using System.Text;
using Rhino;
using Rhino.Commands;

[CommandStyle(Style.ScriptRunner)]
public abstract class RhinoScriptCommand : Command
{
  /// <summary>
  /// Return the command's manifest resource name in the form of
  /// "solution.filename.extension"
  /// </summary>
  public abstract string ResourceName { get; }

  /// <summary>
  /// Return the command's decryption password here
  /// </summary>
  public abstract string Password { get; }

  /// <summary>
  /// Commmand.RunCommand override
  /// </summary>
  protected override Rhino.Commands.Result RunCommand(RhinoDoc doc, RunMode mode)
  {
    Rhino.Commands.Result rc = Rhino.Commands.Result.Success;

    if (!_bIsLoaded)
    {
      string script = ScriptFromResources(ResourceName, Password);
      if (!string.IsNullOrEmpty(script))
      {
        string macro = string.Format("_-RunScript ({0})", script);
        if (RhinoApp.RunScript(macro, false))
          _bIsLoaded = true;
        else
          rc = Result.Failure;
      }
    }

    if (rc == Result.Success)
    {
      string macro = string.Format("_-RunScript ({0})", EnglishName);
      RhinoApp.RunScript(macro, false);
    }

    return rc;
  }

  /// <summary>
  /// Returns a script that was embedded in the assembly's resources.
  /// </summary>
  private string ScriptFromResources(string resourceName, string password)
  {
    if (string.IsNullOrEmpty(resourceName) || string.IsNullOrEmpty(password))
      return null;

    string script = null;

    Assembly assembly = Assembly.GetExecutingAssembly();
    if (null != assembly)
    {
      string encryptedText = null;
      using (Stream stream = assembly.GetManifestResourceStream(resourceName))
      {
        try
        {
          using (StreamReader reader = new StreamReader(stream))
            encryptedText = reader.ReadToEnd();
        }
        catch
        {
        }
      }

      if (!string.IsNullOrEmpty(encryptedText))
        script = Decrypt(encryptedText, password);
    }

    return script;
  }

  /// <summary>
  /// Decrypts a string previously encrypted with RhinoScriptEncrypter.
  /// </summary>
  private string Decrypt(string encryptedText, string password)
  {
    if (string.IsNullOrEmpty(encryptedText) || string.IsNullOrEmpty(password))
      return null;

    string text = null;

    try
    {
      byte[] encryptedBuffer = Convert.FromBase64String(encryptedText);
      byte[] buffer = Decrypt(encryptedBuffer, password);
      text = System.Text.Encoding.Unicode.GetString(buffer);
    }

    catch
    {
      text = null;
    }

    return text;
  }

  /// <summary>
  /// Decrypts a byte array previously encrypted with RhinoScriptEncrypter.
  /// </summary>
  private byte[] Decrypt(byte[] encryptedBuffer, string password)
  {
    if (null == encryptedBuffer || 0 == encryptedBuffer.Length || string.IsNullOrEmpty(password))
      return null;

    byte[] buffer = null;
    CryptoStream cryptoStream = null;

    try
    {
      Rfc2898DeriveBytes secretKey = new Rfc2898DeriveBytes(password, _keySalt);
      RijndaelManaged cipher = new RijndaelManaged();
      ICryptoTransform transform = cipher.CreateDecryptor(secretKey.GetBytes(32), secretKey.GetBytes(16));
      MemoryStream memoryStream = new MemoryStream();
      cryptoStream = new CryptoStream(memoryStream, transform, CryptoStreamMode.Write);
      cryptoStream.Write(encryptedBuffer, 0, encryptedBuffer.Length);
      cryptoStream.FlushFinalBlock();
      buffer = memoryStream.ToArray();
    }

    catch
    {
      buffer = null;
    }

    finally
    {
      if (null != cryptoStream)
        cryptoStream.Close();
    }

    return buffer;
  }

  private byte[] _keySalt = new byte[] { 0x49, 0x76, 0x61, 0x6e, 0x20, 0x4d, 0x65, 0x64, 0x76, 0x65, 0x64, 0x65, 0x76 };
  private bool _bIsLoaded = false;
}

File: RhinoScriptCommand.vb
Imports System
Imports System.IO
Imports System.Security
Imports System.Security.Cryptography
Imports System.Reflection
Imports System.Runtime.InteropServices
Imports System.Text
Imports Rhino
Imports Rhino.Commands

Public MustInherit Class RhinoScriptCommand
  Inherits Rhino.Commands.Command

  ''' <summary>
  ''' Return the command's manifest resource name in the form of
  ''' "solution.filename.extension"
  ''' </summary>
  Public MustOverride ReadOnly Property ResourceName As String

  ''' <summary>
  ''' Return the command's decryption password here
  ''' </summary>
  Public MustOverride ReadOnly Property Password As String

  ''' <summary>
  ''' Decrypts a string previously encrypted with RhinoScriptEncrypter.
  ''' </summary>
  ''' <param name="encryptedText"></param>
  Private Function Decrypt(encryptedText As String, password As String) As String
    If String.IsNullOrEmpty(encryptedText) OrElse String.IsNullOrEmpty(password) Then
      Return Nothing
    End If

    Dim text As String = Nothing

    Try
      Dim encryptedBuffer As Byte() = Convert.FromBase64String(encryptedText)
      Dim buffer As Byte() = Decrypt(encryptedBuffer, password)
      text = System.Text.Encoding.Unicode.GetString(buffer)

    Catch
      text = Nothing
    End Try

    Return text
  End Function

  ''' <summary>
  ''' Commmand.RunCommand override
  ''' </summary>
  Protected Overrides Function RunCommand(doc As RhinoDoc, mode As RunMode) As Rhino.Commands.Result
    Dim rc As Rhino.Commands.Result = Rhino.Commands.Result.Success

    If Not _bIsLoaded Then
      Dim script As String = ScriptFromResources(ResourceName, Password)
      If Not String.IsNullOrEmpty(script) Then
        Dim macro As String = String.Format("_-RunScript ({0})", script)
        If RhinoApp.RunScript(macro, False) Then
          _bIsLoaded = True
        Else
          rc = Result.Failure
        End If
      End If
    End If

    If rc = Result.Success Then
      Dim macro As String = String.Format("_-RunScript ({0})", EnglishName)
      RhinoApp.RunScript(macro, False)
    End If

    Return rc
  End Function

  ''' <summary>
  ''' Returns a script that was embedded in the assembly's resources.
  ''' </summary>
  Private Function ScriptFromResources(resourceName As String, password As String) As String
    If String.IsNullOrEmpty(resourceName) OrElse String.IsNullOrEmpty(password) Then
      Return Nothing
    End If

    Dim script As String = Nothing

    Dim pluginAssembly As Assembly = Assembly.GetExecutingAssembly()
    If pluginAssembly IsNot Nothing Then
      Dim encryptedText As String = Nothing
      Using stream As Stream = pluginAssembly.GetManifestResourceStream(resourceName)
        Try
          Using reader As New StreamReader(stream)
            encryptedText = reader.ReadToEnd()
          End Using
        Catch
        End Try
      End Using

      If Not String.IsNullOrEmpty(encryptedText) Then
        script = Decrypt(encryptedText, password)
      End If
    End If

    Return script
  End Function

  ''' <summary>
  ''' Decrypts a byte array previously encrypted with RhinoScriptEncrypter.
  ''' </summary>
  Private Function Decrypt(encryptedBuffer As Byte(), password As String) As Byte()
    If encryptedBuffer Is Nothing OrElse 0 = encryptedBuffer.Length OrElse String.IsNullOrEmpty(password) Then
      Return Nothing
    End If

    Dim buffer As Byte() = Nothing
    Dim cryptoStream As CryptoStream = Nothing

    Try
      Dim secretKey As New Rfc2898DeriveBytes(password, _keySalt)
      Dim cipher As New RijndaelManaged()
      Dim transform As ICryptoTransform = cipher.CreateDecryptor(secretKey.GetBytes(32), secretKey.GetBytes(16))
      Dim memoryStream As New MemoryStream()
      cryptoStream = New CryptoStream(memoryStream, transform, CryptoStreamMode.Write)
      cryptoStream.Write(encryptedBuffer, 0, encryptedBuffer.Length)
      cryptoStream.FlushFinalBlock()
      buffer = memoryStream.ToArray()

    Catch
      buffer = Nothing
    Finally

      If cryptoStream IsNot Nothing Then
        cryptoStream.Close()
      End If
    End Try

    Return buffer
  End Function

  ''' <summary>
  ''' Private members
  ''' </summary>
  Private _keySalt As Byte() = New Byte() {&H49, &H76, &H61, &H6E, &H20, &H4D, &H65, &H64, &H76, &H65, &H64, &H65, &H76}
  Private _bIsLoaded As Boolean = False

End Class
